File: Cpp/CXR_classify/labelimporter.cpp
#include "labelimporter.h"

#include <iostream>

LabelImporter::LabelImporter(ConfigHandler * configHandler, DatabaseHandler * dbHandler) : Runnable(configHandler, dbHandler)
{

}

void LabelImporter::run()
{
    // std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
    logger->info("Importing label data from CSV");

    std::string elementsJson = configHandler->getColumnsInfoName();
    std::string labelTableName = configHandler->getTableName("label");

    emit attemptUpdateText("Attempting to import image labels");
    dbHandler->addTableToDb(elementsJson, "labels", labelTableName);

    // Open the json with the list of elements we're interested in
    boost::property_tree::ptree columnsJson;
    boost::property_tree::read_json(elementsJson, columnsJson);
    boost::property_tree::ptree elements = columnsJson.get_child("labels");

    // Make the SQL query
    std::string sqlQuery = "COPY " + labelTableName + " (file_name, file_path, ";
    for (boost::property_tree::ptree::value_type & column : elements) {
        if (column.second.get<std::string>("calculation_only") == "false") {
            sqlQuery += (column.first + ",");
        }
    }
    sqlQuery = sqlQuery.substr(0, sqlQuery.length() - 1) + ") FROM '" + configHandler->getCsvPath() + "' DELIMITER ',' CSV HEADER;";

    dbHandler->executeQuery(dbHandler->connection, sqlQuery);

    logger->info("Done importing label data");

    // std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
    // std::cout << "Time difference = " << (std::chrono::duration_cast<std::chrono::nanoseconds> (end - begin).count())/1000000000.0 << "[s]" << std::endl;
    emit attemptUpdateText("Finished importing image labels");
    emit finished();
}

File: Python/DesktopApp/label_importer.py
from stage import Stage, Runnable
import json
import logging
from PyQt5.QtCore import pyqtSlot

class LabelImportStage(Stage):
    def __init__(self, configHandler, dbHandler):
        Stage.__init__(self, configHandler, dbHandler)
        self.labelImporter = self.LabelImporter(configHandler, dbHandler)

    @pyqtSlot()
    def importLabels(self):
        self.threadpool.start(self.labelImporter)

    class LabelImporter(Runnable):
        """Class for importing image labels from CSV."""
        def __init__(self, configHandler, dbHandler):
            Runnable.__init__(self, configHandler, dbHandler)
        
        @pyqtSlot()
        def run(self):
            logging.info("Importing label data from CSV")
            self.signals.attemptUpdateProBarBounds.emit(0,1)
            self.signals.attemptUpdateProBarValue.emit(0)
            self.signals.attemptUpdateText.emit("Importing label data")

            self.dbHandler.add_table_to_db(self.configHandler.getTableName('label'), self.configHandler.getColumnsInfoPath(), 'labels')
            self.import_image_label_data()
            
            self.signals.attemptUpdateProBarValue.emit(1)
            self.signals.attemptUpdateText.emit("Done importing")
            self.signals.finished.emit()
            logging.info("Done importing label data")

        def import_image_label_data(self):
            with open(self.configHandler.getColumnsInfoPath()) as file_reader:
                elements_json = json.load(file_reader)
            elements = elements_json['labels']

            sql_query = 'COPY ' + self.configHandler.getTableName('label') + '(file_name, file_path, '
            for element_name in elements:
                if not elements[element_name]['calculation_only']:
                    sql_query = sql_query + element_name + ','
            sql_query = sql_query[:-1] + ') FROM \'' + self.configHandler.getParentFolder() + "/" + self.configHandler.getCsvPath() + '\' DELIMITER \',\' CSV HEADER;'
            self.dbHandler.executeQuery(self.dbHandler.connection, sql_query)
            self.dbHandler.count_records(self.configHandler.getTableName('label'))
        